File: lab04/ex2/mpi_group.c
/******************************************************************************
* MPI tutorial example code: Groups/Communicators
* FILE: mpi_group.c
* AUTHOR: Blaise Barney
* LAST REVISED:
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#define NPROCS 8

int main(argc,argv)
int argc;
char *argv[];  {
int        rank, new_rank, sendbuf, recvbuf, numtasks,
           ranks1[4]={0,1,2,3}, ranks2[4]={4,5,6,7};
MPI_Group  orig_group, new_group;
MPI_Comm   new_comm;

MPI_Init(&argc,&argv);
MPI_Comm_rank(MPI_COMM_WORLD, &rank);
MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

if (numtasks != NPROCS) {
  printf("Must specify %d. tasks. Terminating.\n",NPROCS);
  MPI_Finalize();
  exit(0);
  }

sendbuf = rank;

/* Extract the original group handle */

=====> Rotina MPI - Identificar o grupo

/* Divide tasks into two distinct groups based upon rank */
if (rank < NPROCS/2) {

=====> Rotina MPI - Define novo grupo com ranks1

  }
else {

=====> Rotina MPI - Define novo grupo com ranks2

  }

/* Create new new communicator and then perform collective communications */

=====> Rotina MPI - Define novo communicator

MPI_Allreduce(&sendbuf, &recvbuf, 1, MPI_INT, MPI_SUM, new_comm);


=====> Rotina MPI - Identifica os processos para os novos grupos

printf("rank= %d newrank= %d recvbuf= %d\n",rank,new_rank,recvbuf);

MPI_Finalize();
}

File: MPI/lab04/ex2/mpi_group.f
C-----------------------------------------------------------------------------
C MPI tutorial example code: Groups/Communicators
C FILE: mpi_group.f
C AUTHOR: Blaise Barney
C LAST REVISED:
C-----------------------------------------------------------------------------
      program group
      include 'mpif.h'

      integer NPROCS
      parameter(NPROCS=8)
      integer rank, new_rank, sendbuf, recvbuf, numtasks
      integer ranks1(4), ranks2(4), ierr
      integer orig_group, new_group, new_comm
      data ranks1 /0, 1, 2, 3/, ranks2 /4, 5, 6, 7/

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

      if (numtasks .ne. NPROCS) then
        print *, 'Must specify ',NPROCS,' tasks. Terminating.'
        call MPI_FINALIZE(ierr)
        stop
      endif

      sendbuf = rank

C     Extract the original group handle

=====> Rotina MPI - Identificar o grupo

C     Divide tasks into two distinct groups based upon rank
      if (rank .lt. NPROCS/2) then

=====> Rotina MPI - Define novo grupo com ranks1

      else 

=====> Rotina MPI - Define novo grupo com ranks2

      endif


=====> Rotina MPI - Define novo communicator

      call MPI_ALLREDUCE(sendbuf, recvbuf, 1, MPI_INTEGER, MPI_SUM,
     &               new_comm, ierr)

=====> Rotina MPI - Identifica os processos para os novos grupos 

      print *, 'rank= ',rank,' newrank= ',new_rank,' recvbuf= ',
     &     recvbuf

      call MPI_FINALIZE(ierr)
      end
